Servers: workaround for boost interprocess

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->
It's an attempt to fix #2409
It seems that on Windows server's shared memory interface can be broken (way too easily IMO) by missing logs.
~~This PR overrides boost's shared memory creation function and uses a fixed location in SC Resource Directory.
**I'm not sure if this is the correct fix**, however I was able to reproduce the bug from #2409 (*) and I can confirm that this PR fixes it. The fix was adopted from https://stackoverflow.com/questions/39742630/boost-shared-memory-cant-be-initialized~~

EDIT: after lengthy tests and back-and-forth with @brianlheim it looks like boost also provides another method for obtaining the stamp for the shm location, enabled by defining `BOOST_INTERPROCESS_BOOTSTAMP_IS_SESSION_MANAGER_BASED`. This seems to fix the issue on my system and seems like the cleanest solution.


(*) I reproduced the bug by going to Event Viewer -> System -> Filter entries by id 6005 -> Clear Log...
## Types of changes

<!-- Delete lines that don't apply -->

- Bug fix

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

brianlheim reviewed (approved):
i'll confess I don't understand why this works or how reliable it is, but if multiple people confirm that it works I think we should merge this. it seems to be a really frustrating and longstanding issue.

dyfer commented:
Thanks @brianlheim 
I definitely would like to get confirmation from more user whether this works well.
The linked SO explains the problem somewhat, I have no information beyond that.

I've been thinking about three more things regarding this:
- does this PR change/touch any public interface of the servers? (I'm thinking of SonicPi etc)
- I haven't really checked where `defaultResourceDirectory` is (haha), does the suggested location seem okay?
- should I include a comment in the source linking to the issue? it is a rather specific fix

brianlheim commented:
ok, I went and actually read the documentation/posts around this and it's actually much clearer than I thought it would be, thanks for linking. What you are doing on your machine reliably reproduces the issue, and looking at the linked discussions and code I don't see any other factors that might affect how this operates on other machines. If you want to get someone else to try it out, though, go ahead!

> does this PR change/touch any public interface of the servers? (I'm thinking of SonicPi etc)

No, it does not. And after reviewing the boost docs I see you are not doing this quite right, so I will redo my review.

> I haven't really checked where defaultResourceDirectory is (haha), does the suggested location seem okay?

Boost.Interprocess's default behavior is to use a directory inside the Common AppData folder, which according to [this page](https://docs.microsoft.com/en-us/windows/win32/shell/csidl) is `C:\Documents and Settings\All Users\Application Data`. I think it's good to use a directory that's under our control, so I don't see a problem with using the default resource directory even if users choose a different one later in sclang. Do you observe that the shared memory resource files placed there are cleaned up after the process exits?

> should I include a comment in the source linking to the issue? it is a rather specific fix

Yes, this should be explained in the source and/or git commit. I will leave it up to you how you want to document it.

brianlheim commented on common/server_shm.hpp:
```diff
@@ -28,6 +28,12 @@
+#ifdef _WIN32
+#    include "SC_Filesystem.hpp"
+#    define BOOST_INTERPROCESS_SHARED_DIR_FUNC                                                                         \
+        void get_shared_dir(std::string& shared_dir) { shared_dir = SC_Filesystem::defaultResourceDirectory + "/shm" }
```
according to boost docs, what you should be doing is

`#define BOOST_INTERPROCESS_SHARED_DIR_FUNC`

prior to including interprocess headers, and then implementing the function

```c++
namespace boost {
    namespace interprocess {
        namespace ipcdetail {
            void get_shared_dir(std::string &shared_dir);
        }
    }
}
```

somewhere (i would do it directly above `namespace detail_server_shm` as you have done here). you would also make that function `inline` since this is a header.

i think it would be safest to set this preprocessor definition in the build system so that it is defined in every file on Windows. there is only one other file in the project that uses B.Interprocess, and it doesn't use these headers, but that may change in the future.

dyfer replied:
> i think it would be safest to set this preprocessor definition in the build system

I don't think I understand. Would the actual function body go into cmake?

brianlheim replied:
sorry, what I'm saying is that this feature is not intended to be used like:

```cpp
#define BOOST_INTERPROCESS_SHARED_DIR_FUNC myFunction() {}
```

but rather

```cpp
// in header
#define BOOST_INTERPROCESS_SHARED_DIR_FUNC
#include <boost/interprocess/_____.hpp>

// in cpp
namespace boost { namespace interprocess { namespace ipcdetail {
void get_shared_dir(std::string &shared_dir) { /* impl */ }
}}}
```

and that setting `BOOST_INTERPROCESS_SHARED_DIR_FUNC` in cmake is slightly more robust than doing it in the header file.

brianlheim commented on common/server_shm.hpp:
```diff
@@ -25,9 +25,21 @@
+#ifdef _WIN32
+namespace boost { namespace interprocess { namespace ipcdetail {
+inline void get_shared_dir(std::string& shared_dir) {
+    shared_dir = SC_Filesystem::instance().getDirectory(SC_Filesystem::DirName::Resource) / "shm";
```
since `shared_dir` is a string, you have to call `.string()` on the right hand side (a `boost::filesystem::path`) to assign to it :)

dyfer commented:
All right, this again now works. I have some questions for consideration:
- I think I understand the point of defining `BOOST_INTERPROCESS_SHARED_DIR_FUNC` in CMake (which I'm not doing ATM) - then we guarantee that it's defined before including the `boost/interprocess/...` headers. But in that case this specific fix is implemented across two relatively distant places: `server_shm.hpp` and `CMakeLists`. What's better? I lean towards keeping the define and the implementation together, as it is now.
- What should be the desired location for the shared memory file? Right now it is _the same folder as the server_, i.e. 
the file is created next to the server executable, wherever SuperCollider is run from. (The file is being deleted when the server quits btw).
  - writing to the executable location seems like not a good policy (Windows programs tended to do that in the Win95-98 era...)
  - we could consider one of SC's `AppData` locations, but we'd need to ensure that the folder exists! 
     - at some point I've tried to do `<current location> + "/shm` and that did not work - I got an information that the directory does not exists
    - I think these are created by the sclang when first started, is that right? If the server is run from the command line, these might not have been created...

brianlheim commented:
for the shared memory directory location, it just needs to be somewhere we can be reliably certain we have permissions to create directories and files. My suggestion would be to use the default user app support directory and an "obvious" subdirectory name like `boost_interprocess_shared_memory` to clarify the purpose for anyone who stumbles across it. we can call `create_directories` to ensure it exists and loudly complain if we lack permissions or some other error occurs.

> I think I understand the point of defining BOOST_INTERPROCESS_SHARED_DIR_FUNC in CMake (which I'm not doing ATM) - then we guarantee that it's defined before including the boost/interprocess/... headers. But in that case this specific fix is implemented across two relatively distant places: server_shm.hpp and CMakeLists. What's better? I lean towards keeping the define and the implementation together, as it is now.

not just that, but so it's defined before including boost/interprocess headers in ALL files, not just this one, which protects against future changes adding/moving things across files. I would view this more as "library configuration" than a localized fix. mixing different configurations of the same library across compilation units is a recipe for undefined behavior/silent disaster, so please do it in CMake.

dyfer commented:
> not just that, but so it's defined before including boost/interprocess headers in ALL files, not just this one,

Yes, that's what I meant. 

Okay, I'll move that definition to cmake.
In that case, should the `ifdef` guards in the header be against `_WIN32`, or rather `BOOST_INTERPROCESS_SHARED_DIR_FUNC` itself?

brianlheim commented:
> In that case, should the ifdef guards in the header be against _WIN32, or rather > BOOST_INTERPROCESS_SHARED_DIR_FUNC itself?

good question; probably the latter, i think this is the condition we really care about

dyfer commented:
I've moved the define to CMakeLists and I've added creating the directory. I'm also checking for the return value from `create_directories`.

However, something is off with that return value. Even though the directory is created, the error is still thrown... What am I doing wrong?

dyfer commented:
Per our discussion, I've changed boost shm location to a non-user directory, since user directories for users with non-ascii names are not supported.
There's still one caveat: if the shm file doesn't get deleted by one user, then scsynth run by another user can't start...

brianlheim commented:
thanks @dyfer !

two questions:

1. did you have a chance to test out that utf8 conversion approach we tried? no worries if not, just wondering.
2. how likely is it that the shm file doesn't get deleted as you say in your last comment?

dyfer commented:
This PR now uses the "session manager" method for obtaining timestamp also provided by boost, which seems to be more reliable. On my system it does fix the linked issue.

brianlheim commented:
the fix is maddeningly simple. :) i'm happy to merge this since it works on your machine. do you want to ask anyone else to test this or are you satisfied with it as it is?

dyfer commented:
> the fix is maddeningly simple. :) 
yeah, especially considering how much of our time it took to arrive at it...

> i'm happy to merge this since it works on your machine. do you want to ask anyone else to test this or are you satisfied with it as it is?
Since this uses a boost-provided functionality and not any new code, I'm inclined to say this should be fine to merge. If if causes any problems, hopefully some Windows-nightly-build users would report back.

Could this be squash merged or do I need to rebase?

brianlheim commented:
sounds good! i will squash-merge this then.

brianlheim reviewed (approved):
thanks again for all your work and patience on this! i have the feeling that we both learned very much and very little at the same time. :)

samaaron commented:
This is amazing work, thank-you so much. I'm happy to throw this into the next Sonic Pi beta and hopefully get a bunch more users trying it out.

Thanks so much again.

dyfer commented:
Thanks @samaaron ! I was actually wondering if this changes anything for Sonic Pi. I thought there were some shared memory related discussions in the past, but maybe I'm misremembering.

samaaron commented:
> Thanks @samaaron ! I was actually wondering if this changes anything for Sonic Pi. I thought there were some shared memory related discussions in the past, but maybe I'm misremembering.

Hopefully this will fix a whole class of bugs we commonly see in the wild. I'm aware of at least one school that has some sort of automated script to futz with the logs to improve the chances of SuperCollider booting on their machines.

Thanks so much.

As an aside, we're now pretty much left with two main classes of SuperCollider-related booting issues:

1. A variety of issues related to how Jack is installed/configured on Linux
2. Issues on Windows where PortAudio seems to struggle with the user's default audio settings and hardware/drivers on Windows.

I completely can see and respect that neither of these are necessarily of direct interest or concern for SuperCollider given that there isn't currently an aim to get things running "out of the box" for all users - rather it's expected that users will have to supply the correct flags on boot to ensure things work. However, if there ever was interest, both could likely be solved within SuperCollider in the future. 1. could be solved by having an alternative backend to Jack - Pulse was discussed briefly and would likely be a good candidate. 2. could at least be alleviated by having an extra flag to scsynth which didn't boot but instead listed all observed audio cards, available drivers, rates and number of ins/outs. That way external systems such as Sonic Pi could use this to generate a list of options to either work through automatically or to present to the user for them to choose from a valid subset - rather than hoping that the defaults will work.

dyfer commented:
> A variety of issues related to how Jack is installed/configured on Linux
> 1. could be solved by having an alternative backend to Jack - Pulse was discussed briefly and would likely be a good candidate. 

Let's see if #5081 moves forward. In the meantime, maybe https://github.com/PortAudio/portaudio/pull/336 gets merged upstream? If that works well, one could just build with PortAudio on Linux.

> Issues on Windows where PortAudio seems to struggle with the user's default audio settings and hardware/drivers on Windows.

Once PortAudio 1.9.17 is released I plan to submit a PR to update it in SC. Let's hope this improves things on Windows...

> could at least be alleviated by having an extra flag to scsynth which didn't boot but instead listed all observed audio cards, available drivers, rates and number of ins/outs

I think it's a good idea. I hope we can consider implementing this at some point.
